sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1005,7 +1006,7 @@ class DataFrame private[sql](
    * @since 1.3.0
    */
   def unionAll(other: DataFrame): DataFrame = withPlan {
-    Union(logicalPlan, other.logicalPlan)
+    CombineUnions(Union(logicalPlan, other.logicalPlan))
```

Comment:
hmm, is it safe to apply optimize rules inside `DataFrame`? A similar case is https://github.com/apache/spark/pull/7920